scripts/hud/cgaz.js

```diff
@@ -922,6 +1292,17 @@ class Cgaz {
 		return angle < 0 ? angle + Math.PI : angle - Math.PI;
 	}
 
+	static rotateVector(vector, angle) {
+		const out = JSON.parse(JSON.stringify(vector));
```

Comment:
This is a hack that shouldn't be used lightly, in this case I'm sure it's faster to do
```js
const out = { x: vector.x, y: vector.y, z: vector.z };
```

but can't this function be simplified to just
```js
const cos = Math.cos(angle);
const sin = Math.sin(angle);

return { 
  x: vector.x * cos - vector.y * sin,
  y: vector.y * cos + vector.x * sin,
  z: vector.z // if there's a z component
}
```

Reply:
yea, tbh I absolutely hate the stringify/parse pattern to get a deep copy of an object. I did a little digging into the history of it and apparently V8 does this very well for small objects. I had entertained the idea of using `structuredClone` just because it's less gross but it's slower for small objects like this.

Your suggestion is the best way forward for this case. I'll make the change to remove the `out` variable.

Reply:
Haha yeah Javascript deep copying is a real rabbit hole. You're right, V8 has some optimisations specifically because the stringify-parse method is so common, I still find it very ugly to see it written though, and crucially only works for a subset of Javascript objects, no functions. Honestly with browsers/node I'd probably always use structuredClone for nested objects as perf may improve in future.

On that - deep copying isn't even needed here. It's only needed for nested objects, these values only have primitive values. So you could've just done a regular shallow copy with `out = { ...vector };`.

Lastly, worth noting that we don't actually have structuredClone. It's part of some web API spec so browsers implement it individually, and Node has it. I've looked into doing it ourselves, even just doing the parse-stringify stuff from C++ which is a bitttttt cleaner, but doing a proper structuredClone implementation would probably be best. Just requires writing an object traversal algorithm (can't remember what structuredClone uses, but I think it's in the spec) and handling all the cursed V8 C++ API object persistence stuff.